config/auth/iap/oauth.go

```diff
@@ -0,0 +1,208 @@
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package config
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"context"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+const (
+	googleTokenURL      = "https://www.googleapis.com/oauth2/v4/token"
+	oauthEmailScope     = "https://www.googleapis.com/auth/userinfo.email"
+	serverStateTokenLen = 60
+)
+
+var (
+	// Configure the default request scopes, this can be overriden
+	oauthScopes = []string{oauthEmailScope}
+)
+
+// Creates a new iapOAuthResponse object
+type iapOAuthResponse struct {
+	AccessToken  string `json:"access_token,omitempty"`
+	RefreshToken string `json:"refresh_token,omitempty"`
+	Type         string `json:"token_type,omitempty"`
+	Expires      int    `json:"expires_in,omitempty"`
+	IDToken      string `json:"id_token,omitempty"`
+}
+
+type oauthReceiver struct {
+	port         int
+	clientState  string
+	killMeNow    bool
+	mainClient   string
+	doneChan     chan error
+	callback     func(*oauth2.Token, *oauth2.Config, string) (string, error)
+	clientId     string
+	clientSecret string
+}
+
+// Cleanly exit the HTTP server.
+func (o *oauthReceiver) killWhenReady(n net.Conn, s http.ConnState) {
+	if o.killMeNow && s.String() == "idle" && o.mainClient == n.RemoteAddr().String() {
+		o.doneChan <- nil
+	}
+}
+
+// NewOAuthConfig creates a new OAuth config
+func (o *oauthReceiver) NewOAuthConfig() *oauth2.Config {
+
+	// return oauth2 Config
+	return &oauth2.Config{
+		ClientID:     o.clientId,
+		ClientSecret: o.clientSecret,
+		RedirectURL:  fmt.Sprintf("http://localhost:%d", o.port),
+		Scopes:       oauthScopes,
+		Endpoint:     google.Endpoint,
+	}
+}
+
+// check the iap's state
+func ValidIAPStateToken(state, clientState string) bool {
+	return state == clientState
+}
+
+// ServeHTTP will serve an HTTP server on localhost used to receive the OAuth callback
+func (o *oauthReceiver) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var tok *oauth2.Token
+	var conf *oauth2.Config
+
+	defer func() {
+		r := recover()
+		if r != nil {
+			log.Printf("panic: %s\n", r)
+			o.doneChan <- fmt.Errorf("%v", r)
+		}
+	}()
+	if r.URL.String() == "/favicon.ico" {
+		return
+	}
+	if r.URL.String() == "/robots.txt" {
+		return
+	}
+
+	o.mainClient = r.RemoteAddr
+
+	conf = o.NewOAuthConfig()
+	oauthCode := r.FormValue("code")
+	state := r.FormValue("state")
+	if oauthCode == "" || state == "" {
+		response := fmt.Errorf("Invalid response from Google's Account Service")
+		o.WebOutput(w, "Uh Oh!", response.Error())
+		o.doneChan <- response
+		return
+	}
+
+	var validateStateFunc func(string, string) bool
```

Comment:
Necessary to save this in a local variable?

Reply:
thanks, have removed this local variable